force_bdss/data_sources/base_data_source_model.py

```diff
@@ -46,43 +46,218 @@ class BaseDataSourceModel(BaseModel):
+        super(BaseDataSourceModel, self).__init__(factory, *args, **kwargs)
 
-        The data source model must have:
-        - input and output slots match between instance and model
-        - all output slots named
-        - no errors in input or output slots
+        if input_slot_info is not None:
+            self.update_slot_info(
+                "input_slot_info", input_slot_info
+            )
 
-        Returns
-        -------
-        errors : list of VerifierErrors
-            The list of all detected errors in the data source model.
-        """
-        try:
-            data_source = self.factory.create_data_source()
-        except Exception:
-            logger.exception(
-                "Unable to create data source from factory '%s', plugin "
-                "'%s'. This might indicate a  programming error",
-                self.factory.id,
-                self.factory.plugin.id,
+        if output_slot_info is not None:
+            self.update_slot_info(
+                "output_slot_info", output_slot_info
             )
-            raise
+
+    # -------------------
+    #      Defaults
+    # -------------------
+
+    def _input_slot_info_default(self):
+        """Default list of InputSlotInfo object, based on length and
+        attributes of Slots tuple returned by slots method on associated
+        BaseDataSource"""
+
+        input_slots, _ = self._generate_slots()
+
+        return [
+            InputSlotInfo(
+                name='', type=slot.type,
+                description=slot.description
+            )
+            for slot in input_slots
+        ]
+
+    def _output_slot_info_default(self):
+        """Default list of OutputSlotInfo object, based on length and
+        attributes of Slots tuple returned by slots method on associated
+        BaseDataSource"""
+        _, output_slots = self._generate_slots()
+
+        return [
+            OutputSlotInfo(
+                name='', type=slot.type,
+                description=slot.description
+            )
+            for slot in output_slots
+        ]
+
+    # -------------------
+    #      Listeners
+    # -------------------
+
+    @on_trait_change("+changes_slots")
+    def _trigger_changes_slots(self, obj, name, new):
+        changes_slots = self.traits()[name].changes_slots
+
+        if changes_slots:
+            self._reset_data_source_slots()
+            self.changes_slots = True
+
+    # -------------------
+    #  Protected Methods
+    # -------------------
+
+    def __getstate__(self):
+        state = pop_dunder_recursive(super().__getstate__())
+        state["input_slot_info"] = [
+            x.__getstate__() for x in self.input_slot_info
+        ]
+        state["output_slot_info"] = [
+            x.__getstate__() for x in self.output_slot_info
+        ]
+
+        return state
+
+    # -------------------
+    #   Private Methods
+    # -------------------
+
+    def _generate_slots(self):
```

Comment:
This method is essentially lifted from the `DataSourceView` class in `force_wfmanager`, it creates an instance of the `BaseDataSource` class assigned to the `factory` trait and uses its `slots` method to generate the expected format of the `input_slot_info` and `output_slot_info` attributes.